GpsDataAnalyzer/calculator/alignment_algorithms.py

```diff
@@ -0,0 +1,279 @@
+from datetime import datetime, timedelta
+import numpy as np
+import time
+import pprint
+
+import utils
+
+def find_lineup(set1, set2):
+  """
+  Optimized algorithm to find the approximate offset between two GPS data sets.
+
+  This algorithm first identifies a primary data set and a secondary one based
+  on which starts later. The offset is then applied to the secondary data set.
+  After mapping the timestamps to the corresponding points from each set, it
+  finds the range that has the least amount of invalid timestamps (has points
+  from only one set or neither) and then finds the optimal offset using the
+  middle of that range as the starting point.
+
+  Args:
+    set1: GpsDataSet object
+    set2: GpsDataSet object
+
+  Returns:
+    Tuple of two datetimes, the start time for set 1 and for set 2 given the
+    calculated offset.
+  """
+
+  # find primary data set (the one with the later start time)
+  set1_start_time = set1.gps_data_list[0].time
+  set2_start_time = set2.gps_data_list[0].time
+  set1_end_time = set1.gps_data_list[-1].time
+  set2_end_time = set2.gps_data_list[-1].time 
+  if set1_start_time > set2_start_time:
+    later_start_time = utils.round_time(set1_start_time)
+    primary_set_index = 0
+  else:
+    later_start_time = utils.round_time(set2_start_time)
+    primary_set_index = 1
+
+  # create dict that maps rounded times to points
+  time_point_mapping = create_time_to_point_mapping(set1, set2)
+
+  span_length = 50
+  range_start = utils.round_time(min(set1_start_time, set2_start_time))
+  range_end = utils.round_time(min(set1_end_time, set2_end_time)) + timedelta(seconds=span_length)
+  total_seconds = int((range_end-range_start).total_seconds())
+
+  best_range_start = range_start
+  lowest_skip_count = span_length
+  previous_range_skip_count = None
+
+  # check each time for how many skipped/invalid entries there are
+  for start_time in [range_start + timedelta(seconds=x) for x in range(total_seconds)]:
```

Comment:
Yeah, I meant to include the range after the end such that it would only check up to span length after the end of the dataset that ends first in line 45, but this is actually double counting that

After more thought I think:
- range start should be  `utils.round_time(min(set1_start_time, set2_start_time)) - timedelta(seconds=span_length//2)`
- range end (aka the last value for start_time) should be `utils.round_time(min(set1_end_time, set2_end_time)) - timedelta(seconds=span_length//2)`
Since I am checking `(-spanlength//2, spanlength//2)` from whatever spot has the most overlapping